### PR TITLE
Add support for deploying application from CI server

### PR DESCRIPTION
**Pull request by ido-ran**
I use Jenkins as CI server. As I move to AppHarbor I need a way to automate the deployment of my web-application to AppHarbor.
AppHarbor CLI was designed to be used by human and this change allow machine to use it too.

There are several changes:
1. I've add to the DeployAppCommand two parameters to accept the username and password from the console. This  allow to run the command with apriori login manually. It also does not store the login token and allow two or more deployment at the same time on the same machine with the same user (the way CI usually works).
1. I've added some print outs to allow me to understand what happen by reading the CI log.
2. The source-directory parameter did not take into account the fact that the Tar library use the current working directory as a reference when creating the tar entries. This cause the resulting Tar file to has the source-directory prefixed to all the file names which result in bad deployment in AppHarbor. For example, if you run the command from c:\deploy and you have the file c:\deploy\a\b\image.png and you use source-directory=a\b the result tar file when extracted to c:\temp will be c:\temp\a\b\image.png instead of c:\temp\image.png.
3. I bump the version number so it will be easy to distinguish between old and new exe files.
4. Last but not least, during the executing of deploy app command you show progress bar in the console. This does not work well in CI servers which redirect the output and have no console window.
   For now I've create a progress bar factory that check if a console window exist and either create the real console progress bar or a "null" NOP one. I didn't use DI because the progress bar was instantiated in a static method as well as inside the command. A better solution will be to register the correct progress bar implementation when the application starts.

I think it worth spending the time to separate the actual work of deploying from the informative output. I though of creating a "status update" interface for the deploy command and pass it in as a dependency. This way we could decide how to implement the update of which user is been used and how to implement the progress bar.
That's been said, this is a working version that I am currently use in my Jenkins server.

### CODE REVIEW

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 1]
```diff
@@ -15,38 +15,74 @@ public class DeployAppCommand : ApplicationCommand
 	{
 		private string _message;
 		private DirectoryInfo _sourceDirectory;
+		private bool _isDeployFromCurrentDirectory;
```
Comment (runesoerensen) [inline src/AppHarbor/Commands/DeployAppCommand.cs:18]: This seems unnecessary - what value does using this field add over the existing implementation?

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 2]
```diff
@@ -15,38 +15,74 @@ public class DeployAppCommand : ApplicationCommand
...
+		private string _username;
```
Comment (runesoerensen) [inline src/AppHarbor/Commands/DeployAppCommand.cs:19]: Sign in is already handled by the "login" command and isn't really related specifically to deploying an app

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 3]
```diff
@@ -15,38 +15,74 @@ public class DeployAppCommand : ApplicationCommand
...
 
 			var uploadCredentials = GetCredentials();
 
+			if (_isDeployFromCurrentDirectory)
+			{
+				_writer.WriteLine("Deploying from current directory.");
+			}
+			else
+			{
+				_writer.WriteLine("Deploying from " + _sourceDirectory.FullName);
```
Comment (runesoerensen) [inline src/AppHarbor/Commands/DeployAppCommand.cs:82]: It's a good idea to add this information, but we might as well just print the entire directory every time (and avoid the if/else block)

Comment (runesoerensen) [inline src/AppHarbor/Commands/DeployAppCommand.cs:82]: Also use `string.Format("Deploying from \"{0}\"", _sourceDirectory.FullName)` instead

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 4]
```diff
@@ -67,13 +103,13 @@ protected override void InnerExecute(string[] arguments)
 						Timeout = (int)TimeSpan.FromHours(2).TotalMilliseconds,
 					};
 
-					var progressBar = new MegaByteProgressBar();
+					ProgressBarPresenter progressBar = ProgressBarFactory.CreateMegaByteProgressBar();
```
Comment (runesoerensen) [inline src/AppHarbor/Commands/DeployAppCommand.cs:106]: It'd be great if we could use dependency injection for this instead of a static factory

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 5]
```diff
@@ -67,13 +103,13 @@ protected override void InnerExecute(string[] arguments)
...
 					request.UploadProgressEvent += (object x, UploadProgressArgs y) => progressBar
 						.Update("Uploading package", y.TransferredBytes, y.TotalBytes);
 
 					transferUtility.Upload(request);
 
-					Console.CursorTop++;
```
Comment (runesoerensen) [inline src/AppHarbor/Commands/DeployAppCommand.cs:76]: Doesn't this break the existing functionality when using it in the CLI?

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 6]
```diff
@@ -133,5 +169,27 @@ private void TriggerAppHarborBuild(FederatedUploadCredentials credentials)
 				}
 			}
 		}
+
```
Comment (runesoerensen) [inline src/AppHarbor/Commands/DeployAppCommand.cs:172]: Remove this part and use the login command instead

File: src/AppHarbor/Commands/LoginUserCommand.cs [hunk 7]
```diff
@@ -41,22 +41,8 @@ protected override void InnerExecute(string[] arguments)
...
-
-			if (accessToken == null)
-			{
-				throw new CommandException("Couldn't log in. Try again");
-			}
-
-			return accessToken;
+			return AccessTokenHelper.GetAccessToken(username, password);
 		}
+
```
Comment (runesoerensen) [inline src/AppHarbor/Commands/LoginUserCommand.cs:46]: Remove whitespace

File: src/AppHarbor/NullProgressBar.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
+﻿
```
Comment (runesoerensen) [inline src/AppHarbor/NullProgressBar.cs:1]: Remove whitespace

File: src/AppHarbor/NullProgressBar.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+namespace AppHarbor
+{
+	/// <summary>
+	/// Reperesent a null progress bar that does nothing.
```
Comment (runesoerensen) [inline src/AppHarbor/NullProgressBar.cs:5]: Remove comment (it's evident from the class name already)

File: src/AppHarbor/ProgressBarFactory.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
+﻿using System;
+using System.Diagnostics;
+
+namespace AppHarbor
+{
+	/// <summary>
```
Comment (runesoerensen) [inline src/AppHarbor/ProgressBarFactory.cs:6]: Remove comment (it's evident from the class name already)

File: src/AppHarbor/ProgressBarPresenter.cs [hunk 11]
```diff
@@ -0,0 +1,16 @@
+﻿namespace AppHarbor
+{
+    /// <summary>
```
Comment (runesoerensen) [inline src/AppHarbor/ProgressBarPresenter.cs:3]: Remove comments

File: src/AppHarbor/ProgressBarFactory.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+			get
+			{
+				if (hasConsoleWindow == null)
+				{
+					try
+					{
+						int w = Console.BufferWidth;
+						hasConsoleWindow = true;
+					}
+					catch (Exception)
```
Comment (runesoerensen) [inline src/AppHarbor/ProgressBarFactory.cs:36]: It's an `ArgumentOutOfRangeException` we want to catch, right?

File: src/AppHarbor/ProgressBarFactory.cs [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace AppHarbor
+{
+	/// <summary>
+	/// Factory for creating ProgressBar that match the current environment
+	/// in which the application runs.
+	/// </summary>
+	public static class ProgressBarFactory
+	{
+		private static Nullable<bool> hasConsoleWindow;
```
Comment (runesoerensen) [inline src/AppHarbor/ProgressBarFactory.cs:12]: Why is this necessary? Can't `HasConsoleWindow` just return null or false?

Since this will only be called once it seems unnecessary to cache the value and adds complexity. You might even be better off merging  the method that returns a progress bar with the logic in `HasConsoleWindow`

File: src/AppHarbor/ProgressBarFactory.cs [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+		public static ProgressBarPresenter CreateMegaByteProgressBar()
```
Comment (runesoerensen) [inline src/AppHarbor/ProgressBarFactory.cs:13]: The naming is inaccurate - `CreateProgressBar` is probably better (when the "Presenter" stuff has been renamed)

Comment (runesoerensen) [inline src/AppHarbor/ProgressBarFactory.cs:13]: Alternatively just `Create` since the class name is `ProgressBarFactory`

File: src/AppHarbor/Commands/DeployAppCommand.cs (hunk 1 above)
Comment (ido-ran) [inline src/AppHarbor/Commands/DeployAppCommand.cs:18]: This is mostly for writing information so it will be clear where the tool took the deployment files

File: src/AppHarbor/ProgressBarPresenter.cs [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+    /// Represent a presenter for progress bar information.
+    /// </summary>
+    public interface ProgressBarPresenter
```
Comment (runesoerensen) [inline src/AppHarbor/ProgressBarPresenter.cs:6]: Maybe just remove the "Presenter" semantics?

Comment (runesoerensen) [inline src/AppHarbor/ProgressBarPresenter.cs:6]: Rename to `IProgressBar`

File: src/AppHarbor/Commands/DeployAppCommand.cs (hunk 2 above)
Comment (ido-ran) [inline src/AppHarbor/Commands/DeployAppCommand.cs:19]: I know but the whole point of this change was to allow a single command call to deploy the application.
When I use the tool in CI server I can't run the login command manually.
I made it so it the username parameter is not specified to normal logged in user is used so it is compatibly with the exist usage pattern.

File: src/AppHarbor/Commands/DeployAppCommand.cs (hunk 3 above)
Comment (ido-ran) [inline src/AppHarbor/Commands/DeployAppCommand.cs:82]: I agree

File: src/AppHarbor/Commands/DeployAppCommand.cs (hunk 4 above)
Comment (ido-ran) [inline src/AppHarbor/Commands/DeployAppCommand.cs:106]: I agree. I'll make the changes to use the DI

File: src/AppHarbor/Commands/DeployAppCommand.cs (hunk 5 above)
Comment (ido-ran) [inline src/AppHarbor/Commands/DeployAppCommand.cs:76]: As far as I understood it this actually equals to calling WriteLine. Am I wrong?

File: src/AppHarbor/Commands/DeployAppCommand.cs (hunk 6 above)
Comment (ido-ran) [inline src/AppHarbor/Commands/DeployAppCommand.cs:172]: but that was the whole idea here, to not require run the login command before deploying so CI tools can use AppHarbor.exe

File: src/AppHarbor/Commands/LoginUserCommand.cs (hunk 7 above)
Comment (ido-ran) [inline src/AppHarbor/Commands/LoginUserCommand.cs:46]: np :+1:

File: src/AppHarbor/ProgressBarFactory.cs (hunk 13 above)
Comment (ido-ran) [inline src/AppHarbor/ProgressBarFactory.cs:12]: OK. I can remove the caching, you right about that.
I don't like return null because then I have to check all the time.
Also I do like to turn NullProgressBar into ForwardOnlyProgressBar which does write progress only in forward only way, like
10%
20%
and so on once every 1-3 seconds so it will not clutter the log.

**Comment by runesoerensen** [comment]
Thanks for this pull request. It seems cleaner to simply execute the "login" command prior to the "deploy" command, rather than allowing login as part of the `DeployAppCommand. The parameters you've added to the deploy command can just be moved to the login command. This way you can still sign in with command line parameters prior to each deployment.

It'd be great if you could separate this into three pull requests that are relevant for each "logical" change:
1. Adds the new username/password parameters to the LoginUserCommand.
2. Makes the progress bar compatible with situations where there is no console window.
3. Fixes the source directory/tarball issue you mention and adds the new output showing which directory is being deployed.

File: src/AppHarbor/ProgressBarFactory.cs (hunk 12 above)
Comment (ido-ran) [inline src/AppHarbor/ProgressBarFactory.cs:36]: No, actually it is `IOException`
And you are right, it's better to be specific about exceptions

File: src/AppHarbor/ProgressBarPresenter.cs (hunk 11 above)
Comment (ido-ran) [inline src/AppHarbor/ProgressBarPresenter.cs:3]: You really don't like comments :)
np

File: src/AppHarbor/ProgressBarFactory.cs (hunk 14 above)
Comment (ido-ran) [inline src/AppHarbor/ProgressBarFactory.cs:13]: I thought to remain consistent with the class name MegaByteProgressBar.
In case you will add GigaByteProcessBar.

File: src/AppHarbor/ProgressBarPresenter.cs (hunk 15 above)
Comment (ido-ran) [inline src/AppHarbor/ProgressBarPresenter.cs:6]: I'll do that.
In my project I don't use the I for interfaces because I used to also use Java so I took their side

**Comment by ido-ran** [comment]
It's not that good actually.
When using tools in CI server it dangerous to rely on environment variables that might change any time.
For example, we plan to deploy to staging application and to our production from the same CI server.
If we will use different users, to make sure we don't accidentally deploy to production, it might cause a race condition between two concurrently running deploy jobs.

I understand your need to keep the CLI tool simple. What do you think about adding a base class to add the commands that require login so you'll be able to specify -username and -password to all of them? I know it's the opposite direction you were talking, but still, what do you think?

File: src/AppHarbor/Commands/DeployAppCommand.cs (hunk 1 above)
Comment (runesoerensen) [inline src/AppHarbor/Commands/DeployAppCommand.cs:18]: Ah alright.. I suppose https://github.com/appharbor/appharbor-cli/pull/14#discussion_r5110268 changes that

File: src/AppHarbor/Commands/DeployAppCommand.cs (hunk 2 above)
Comment (runesoerensen) [inline src/AppHarbor/Commands/DeployAppCommand.cs:19]: Alright got it.. At any rate this is not the right place to put this logic, the question is where it should be then. I'll follow up in the related comment

**Comment by ido-ran** [comment]
I'm closing this pull request per the discussion we had to separate the CI compatible command to "ci-deploy" command.
